src/proxyConnect.ts

```diff
@@ -0,0 +1,105 @@
+import { Socket } from 'net'
+import { DOMParser } from 'xmldom'
+import { EventEmitter } from 'events'
+import { decode } from 'iconv-lite'
+import { ENCODING } from './xdebugConnection'
+
+export const DEFAULTIDEKEY = 'vsc'
+
+/** Informs proxy of incoming connection and who to pass data back to. */
+export class ProxyConnect extends EventEmitter {
+    /** Port editor is listening on (default: 9001 */
+    private _port: number
+    /** a CLI binary boolean option (default: 1) */
+    private _allowMultipleSessions: number
+    /** host domain or ip (default: 127.0.0.1) */
+    private _host: string
+    /** unique key that allows the proxy to match requests to your editor. (default: DEFAULTIDEKEY) */
+    private _key: string
+    /** proxy response data parser */
+    private _parser = new DOMParser()
+    /** tcp connection to communicate with proxy server */
+    public _socket = new Socket()
+    /** milliseconds to wait before giving up */
+    private _timeout: number
+    public msgs: Map<string, string>
+    private _isRegistered = false
+    private _resolve: Function
+
+    constructor(host = '127.0.0.1', port = 9001, allowMultipleSessions = 1, key = DEFAULTIDEKEY, timeout = 3000) {
+        super()
+        this._allowMultipleSessions = allowMultipleSessions
+        this._host = host
+        this._key = key
+        this._port = port
+        this._timeout = timeout
+        this.msgs = new Map([
```

Comment:
Using a `Map` with string identifiers for this means it is not typesafe. When these are only used once, I would inline these. If they are needed multiple times, define a method that returns it.

Reply:
How is it not typesafe? In its declaration, on line 25: `public msgs: Map<string, string>`, it specifies type.
A method would justifiable if the `msgs` had a way to change but there isn't and they don't. 
Using a map allowed us to group their intended functionality whilst reducing variable name verbosity.

Reply:
When calling msgs.get() you could easily have a typo in the key. You also get no autocompletion for the keys.

An object would also be fine.

Reply:
```typescript
    public msgs: {
        defaultError: 'Unknown Error',
        deregisterInfo: string,
        deregisterSuccess: 'Deregistration successful',
        duplicateKey: 'IDE Key already exists',
        nonexistentKey: 'No IDE key',
        registerInfo: string,
        registerSuccess: 'Registration successful',
        resolve: string,
        timeout: string,
    }

    constructor() {
        ...
        this.msgs.deregisterInfo = `Deregistering ${this._key} with proxy @ ${this._host}:${this._port}`
        this.msgs.registerInfo = `Registering ${this._key} with proxy @ ${this._host}:${this._port}`
        this.msgs.resolve = `Failure to resolve ${this._host}`
        this.msgs.timeout = `Timeout connecting to ${this._host}:${this._port}`
        ...
    }
```
yes?

Reply:
Oh, I believe this would be more suitable

```typescript
    
    public msgs:{ [index:string] : {message: string} } = {};
    constructor() {
        ...
        this.msgs = {
            defaultError: 'Unknown Error',
            deregisterInfo: `Deregistering ${this._key} with proxy @ ${this._host}:${this._port}`,
            deregisterSuccess: 'Deregistration successful',
            duplicateKey: 'IDE Key already exists',
            nonexistentKey: 'No IDE key',
            registerInfo: `Registering ${this._key} with proxy @ ${this._host}:${this._port}`,
            registerSuccess: 'Registration successful',
            resolve: `Failure to resolve ${this._host}`,
            timeout: `Timeout connecting to ${this._host}:${this._port}`,
        }
        ...
    }
```
yes?

Reply:
No, because the `key` is still `string`. But if you move the object from the constructor to the initialiser instead and let the type be inferred, it would be safe.

Reply:
@felixfbecker included an interface defining `ProxyMessages` to ensure the key is type safe.